File: cmd/ih.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"regexp"
	"strconv"
	"strings"

	"github.com/merj/gokogiri"
)

func usage() {
	fmt.Fprintf(os.Stderr, "usage: %s n\n", path.Base(os.Args[0]))
	os.Exit(1)
}

func main() {
	if len(os.Args) != 2 {
		usage()
	}

	n, err := strconv.Atoi(os.Args[1])
	if err != nil {
		os.Exit(1)
	}

	re := regexp.MustCompile(`[\s\W[:space:]]+`)

	for i := 0; i < n; i++ {
		buf, err := ioutil.ReadFile(fmt.Sprintf("%d.html", i))
		if err != nil {
			fmt.Fprintf(os.Stderr, "%s\n", err)
			continue
		}
		doc, err := gokogiri.ParseHtml(buf)
		if err != nil {
			fmt.Fprintf(os.Stderr, "%s\n", err)
			continue
		}
		for _, blacklist := range []string{"title", "script", "style"} {
			nodeRm, _ := doc.Search(fmt.Sprintf("//%s/text()", blacklist))
			for j := range nodeRm {
				nodeRm[j].Remove()
			}
		}
		ns, err := doc.Search("//text()")
		if err != nil {
			fmt.Fprintf(os.Stderr, "%s\n", err)
			continue
		}
		var s string
		for _, n := range ns {
			s += n.String()
		}
		doc.Free()

		s = gsub(re, s)
		s = strings.TrimSpace(s)
		s = strings.ToLower(s)
		ioutil.WriteFile(fmt.Sprintf("%d.txt", i), []byte(s), 0777)
	}
}

func gsub(re *regexp.Regexp, str string) string {
	s := ""
	i := 0
	for _, v := range re.FindAllSubmatchIndex([]byte(str), -1) {
		g := []string{}
		for i := 0; i < len(v); i += 2 {
			g = append(g, str[v[i]:v[i+1]])
		}
		s += str[i:v[0]] + " "
		i = v[1]
	}
	return s + str[i:]
}

File: cmd/ih.rb
#!/usr/bin/env ruby

require 'nokogiri'

if ARGV.length != 1
  STDIN.puts("usage: #{File.basename(__FILE__)} n")
  exit(1)
end

n = ARGV[0].to_i()

BLACKLIST = %w[title script style]

i = 0
while i < n
	f = File.open("#{i}.html", 'r')
	h = Nokogiri::HTML(f)
	f.close()
    ns = h.search('//text()')
    BLACKLIST.each do |t|
      ns -= h.search("//#{t}/text()")
    end
    s = ns.text
    s.scrub!
    s.gsub!(/[\s\W[:space:]]+/, ' ')
    ts = s.split(' ')
    ts.select! do |t|
      t.length >= 2 or t.length <= 25
    end
    s = ts.join(' ')
    s.downcase!()
    IO.write("#{i}.txt", s)
	i += 1
end
